File: codejam/2014/1C/b.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int M = 1000000007;

int perms(int i, char d[200], char last, char **cc) {
  if (i == 0) {
    return 1;
  }
  int shift = 0;
  int count = 0;
  while (i >> shift != 0) {
    if (((i >> shift) & 1) == 1) {
      char *c = cc[shift];
      char dcpy[200];
      memcpy(dcpy, d, sizeof(dcpy));
      char usable = 1;
      char nlast = last;
      if (dcpy[nlast] == 0 || nlast == c[0]) {
        for (int j = 0; c[j] != 0; j++) {
          char cur = c[j];
          if (cur != nlast && dcpy[nlast] != 0) {
            usable = 0;
          }
          dcpy[cur]--;
          nlast = cur;
        }
      }
      else {
        usable = 0;
      }
      if (usable) {
        count += perms(i & (~(1 << shift)), dcpy, nlast, cc);
        count %= M;
      }
    }
    shift++;
  }
  return count;
}

int main(void) {
  int T;
  scanf("%d", &T);
  for (int t = 1; t <= T; t++) {
    int N;
    scanf("%d", &N);
    char** cc = malloc(sizeof(char*) * N);
    for (int i = 0; i < N; i++) {
      char* c = malloc(sizeof(char) * 200);
      scanf("%s", c);
      cc[i] = c;
    }
    char d[200] = {0};
    for (int i = 0; i < N; i++) {
      char* c = cc[i];
      for (int j = 0; c[j] != 0; j++) {
        d[c[j]]++;
      }
    }
    printf("Case #%d: %d\n", t, perms((1 << N) - 1, d, 0, cc));
    for (int i = 0; i < N; i++) {
      free(cc[i]);
    }
    free(cc);
  }
  return 0;
}

File: codejam/2014/1C/B.jl
#!/usr/bin/env julia

M = 1_000_000_007

T = int(readline(STDIN))

function perms(i, d, last, cars)
    if i == 0
        return 1
    end
    shift = 0
    count = 0
    while i >> shift != 0
        if (i >> shift) & 1 == 1
            car = cars[shift+1]
            dcpy = copy(d)
            usable = true
            nlast = last
            if get(dcpy, nlast, 0) == 0 || nlast == car[1]
                for j in 1:endof(car)
                    cur = car[j]
                    if cur != nlast && get(dcpy, nlast, 0) != 0
                        usable = false
                    end
                    dcpy[cur] -= 1
                    nlast = cur
                end
            else
                usable = false
            end
            if usable
                count += perms(i & (~(1 << shift)), dcpy,
                               car[end], cars)
                count %= M
            end
        end
        shift += 1
    end
    count
end

for t in 1:T
    N = int(readline(STDIN))
    cars = split(readline(STDIN))
    mask = (1 << N) - 1
    d = Dict()
    for i in 1:N
        for j in 1:endof(cars[i])
            d[cars[i][j]] = get(d, cars[i][j], 0) + 1
        end
    end
    @printf("Case #%d: %d\n", t, perms(mask, d, nothing, cars));
end
